spec/models/course/assessment/submission_spec.rb

```diff
@@ -572,6 +572,51 @@ def unsubmit_and_save_subject
       end
     end
 
+    describe '#resubmit_programming!' do
+      let(:assessment_traits) { [:with_all_question_types, :autograded] }
+      let!(:earlier_answer) do
+        answer = create(:course_assessment_answer_multiple_response, :graded,
+                        assessment: assessment,
+                        question: assessment.multiple_response_questions.first.acting_as,
+                        submission: submission1, creator: user1).acting_as
+        answer.update_column(:created_at, 1.day.ago)
+        submission1.reload
+        # Submission creates a grading job which could make answers go to the graded state, need to
+        # wait for them to finish.
+        wait_for_job
+        answer
+      end
+
+      subject { submission1 }
+
+      context 'when the submission is published' do
+        let(:submission1_traits) { :published }
+        before do
+          submission1.points_awarded = 200
+          submission1.save!
+        end
+
+        it 'resets experience points, publish_at and publisher attributes, but not submitted_at' do
+          expect(subject.points_awarded).not_to be_nil
+          expect(subject.submitted_at).not_to be_nil
+          original_submitted_at = subject.submitted_at
+          expect(subject.published_at).not_to be_nil
+          expect(subject.publisher).not_to be_nil
+          subject.resubmit_programming!
+          expect(subject.points_awarded).to be_nil
+          expect(subject.submitted_at).to equal(original_submitted_at)
+          expect(subject.published_at).to be_nil
+          expect(subject.publisher).to be_nil
+        end
+
+        it 'sets all current programming answers in the submission to submitted' do
+          subject.resubmit_programming!
+          expect(subject.current_programming_answers.all?(&:submitted?)).to be(true)
+          expect(earlier_answer.reload).to be_graded
```

Comment:
`earlier_answer` is not a current_answer so it would still remain in the `graded` state even if `resubmit_programming` accidentally unsubmits all current answers.